File: python/permutations.py
class Solution:
    def permute(self, nums: list[int]) -> list[list[int]]:
        """time complexity: O(n! * n)"""
        result: list[list[int]] = []
        subset: list[int] = []
        pick: list[bool] = [False] * len(nums)
        self.dfs(subset, result, pick, nums)
        return result

    def dfs(
        self,
        subset: list[int],
        result: list[list[int]],
        pick: list[bool],
        nums: list[int],
    ) -> None:
        if len(subset) == len(nums):
            result.append(subset.copy())
            return
        for i in range(len(nums)):
            if pick[i]:
                continue
            subset.append(nums[i])
            pick[i] = True
            self.dfs(subset, result, pick, nums)
            subset.pop()
            pick[i] = False


def test_permute_case_1():
    # arrange
    nums: list[int] = [1, 2, 3]
    expected: list[list[int]] = [
        [1, 2, 3],
        [1, 3, 2],
        [2, 1, 3],
        [2, 3, 1],
        [3, 1, 2],
        [3, 2, 1],
    ]

    # act
    solution = Solution()
    actual = solution.permute(nums)

    # assert
    expected.sort()
    actual.sort()
    assert expected == actual


def test_permute_case_2():
    # arrange
    nums: list[int] = [0, 1]
    expected: list[list[int]] = [[0, 1], [1, 0]]

    # act
    solution = Solution()
    actual = solution.permute(nums)

    # assert
    expected.sort()
    actual.sort()
    assert expected == actual


def test_permute_case_3():
    # arrange
    nums: list[int] = [1]
    expected: list[list[int]] = [[1]]

    # act
    solution = Solution()
    actual = solution.permute(nums)

    # assert
    expected.sort()
    actual.sort()
    assert expected == actual

File: rust/src/code/permutations.rs
use std::collections::VecDeque;

struct Solution {}

impl Solution {
    pub fn permute(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut result: Vec<Vec<i32>> = Vec::new();
        if nums.len() == 1 {
            result.push(nums.clone());
            return result;
        }

        let mut ns = VecDeque::from(nums);
        for _ in 0..ns.len() {
            if let Some(n) = ns.pop_front() {
                let mut perms = Self::permute(ns.clone().into_iter().collect::<Vec<i32>>());

                for perm in perms.iter_mut() {
                    perm.push(n)
                }
                result.append(&mut perms);
                ns.push_back(n)
            }
        }

        result
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn permute_case_1() {
        // arrange
        let nums: Vec<i32> = vec![1, 2, 3];
        let mut expected: Vec<Vec<i32>> = vec![
            vec![1, 2, 3],
            vec![1, 3, 2],
            vec![2, 1, 3],
            vec![2, 3, 1],
            vec![3, 1, 2],
            vec![3, 2, 1],
        ];

        // act
        let mut actual = Solution::permute(nums);

        // arrange
        expected.sort();
        actual.sort();
        assert_eq!(expected, actual);
    }

    #[test]
    fn permute_case_2() {
        // arrange
        let nums: Vec<i32> = vec![0, 1];
        let mut expected: Vec<Vec<i32>> = vec![vec![0, 1], vec![1, 0]];

        // act
        let mut actual = Solution::permute(nums);

        // arrange
        expected.sort();
        actual.sort();
        assert_eq!(expected, actual);
    }

    #[test]
    fn permute_case_3() {
        // arrange
        let nums: Vec<i32> = vec![1];
        let mut expected: Vec<Vec<i32>> = vec![vec![1]];

        // act
        let mut actual = Solution::permute(nums);

        // arrange
        expected.sort();
        actual.sort();
        assert_eq!(expected, actual);
    }
}
